### ISSUE TITLE
Merging mangles objects like FormData and File

### ISSUE THREAD

**Issue by eltiare** [KyleAMathews/deepmerge#90]
It seems like FormData and File are converted to empty objects in the resulting merged set. 

```
import merge from 'deepmerge';
let defaults = {};
let incoming = { formData: new FormData() };
merge(defaults, incoming); // => { formData: {} }
```

**Comment by jcoc611** [comment]
It even messes up built-in objects like `Set`s:

```javascript
import merge from 'deepmerge';
merge({}, {s: new Set()}); // => { s: {} }
```

The issue is in the dependency [is-mergeable-object](https://github.com/TehShrike/is-mergeable-object). I suggest we use [is-plain-object](https://github.com/jonschlinkert/is-plain-object) instead. It would correctly handle these cases.

```javascript
import isMergeable from 'is-mergeable-object';
import isPlain from 'is-plain-object';
isMergeable(new Set()); // true
isPlain(new Set()); // false
```

**Comment by TehShrike** [comment]
Is it even possible to create cloned instances of `FormData` or `File`?

Are you two suggesting that `FormData`, `File`, and `Set` should be copied to the target, instead of being cloned?

Switching from `is-mergeable-object` to `is-plain-object` would be switching from blacklist ("merge everything except for these special types") to whitelist ("only merge this limited definition of objects").

It would be a big change.  Maybe it would be preferable?

You may be interested in [`merge-deep`](https://github.com/jonschlinkert/merge-deep), which is based on `is-plain-object`.

Anyone else have any thoughts?

**Comment by eltiare** [comment]
I don't think it's possible to clone them. I think a white list would make more sense. Copy those that can't be cloned/merged and clone all the others. It would be the least surprising. Maybe caution people to keep in mind that it's copied for all unclonables.

**Comment by jcoc611** [comment]
I think its a matter of preference. Depends on the use case of the library. If we want people to use the library to clone arbitrary JavaScript objects, then the library should somehow merge things like `Set`, `Map`, as well as custom user objects like `FormData`. It could happen by reference (the original object is exposed), or we could somehow try to clone the object using its prototype. [This package](https://github.com/pvorb/clone) implements this type of cloning, but I'd have to test it to see if it can clone any object.

A different approach could be to allow users to define merge strategies per object type. For example, if a user wants to merge instances of `Set` by a set union operation, they could do the following:

```javascript
merge(x, y, {
    using: {
        [Set]: (a, b) => {
            let union = new Set(a)
            for (let elem of b) {
                union.add(elem)
            }
            return union
        }
    }
})
```

Otherwise, if we just want to have the library handle plain objects (i.e. like the ones allowed in JSON), then it's perfectly fine as it is right now.

**Comment by TehShrike** [comment]
If anyone favors switching to the whitelist-based approach - is there any reason you wouldn't just use [`merge-deep`](https://github.com/jonschlinkert/merge-deep) instead of this library?

**Comment by eltiare** [comment]
If the other library works I don't have a problem with doing so, yours is just the one I settled on for whatever reason and figured I'd report the issue.

**Comment by eltiare** [comment]
If you keep it as is you'll probably want to include a warning in the readme that this is only for plain objects.

**Comment by eltiare** [comment]
You're fundamentally changing the objects and the output is just plain objects though, right?

**Comment by TehShrike** [comment]
A more accurate way to think of it is: this library (currently) only copies the properties of objects.  It doesn't do anything with the objects themselves.

**Comment by eltiare** [comment]
If you decide not to change the library, that sounds like a fair warning: the output will only have the properties of the objects.

**Comment by TehShrike** [comment]
This would be fixed by #96.  Check out the documentation changes there, see if you think it looks reasonable.
